### PR TITLE
refactor: FTP Connection pooling

### PR DESCRIPTION
**Pull request by slavek-kucera**

### CODE REVIEW

File: clients/vscode-hlasmplugin/src/connectionPool.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    async withClient<R>(action: (c: T) => R | PromiseLike<R>): Promise<R> {
+        return this.semaphore.locked(async () => {
+            const client = await this.getClient();
+            const gen = this.generation;
+            try {
+                const result = await Promise.resolve(action(client));
+
+                if (gen === this.generation && !this.disposed && this.clientDetails.reusable(client)) {
```
Comment (VitGottwald) [inline clients/vscode-hlasmplugin/src/connectionPool.ts:77]: How could `gen !== this.generation`  when on line 73 we have `const gen = this.generation` ?

Comment (slavek-kucera) [reply clients/vscode-hlasmplugin/src/connectionPool.ts:77]: When somebody calls `closeClients`.

Comment (VitGottwald) [reply clients/vscode-hlasmplugin/src/connectionPool.ts:77]: I see now, overlooked the `await` on line 75. ✅

File: clients/vscode-hlasmplugin/src/hlasmExternalFilesFtp.ts [hunk 2]
```diff
@@ -47,34 +48,30 @@ class DatasetUriDetails implements ClientUriDetails {
...
     private stateChanged = new vscode.EventEmitter<boolean>();
 
-    get onStateChange() {
-        return this.stateChanged.event;
+    constructor(private context?: vscode.ExtensionContext) {
+        this.clientPool = new ConnectionPool<ftp.Client>({
+            create: () => this.getConnectedClient(),
+            reusable: (client: ftp.Client) => !client.closed && !this.clientSuspended,
+            close: (client: ftp.Client) => { try { client.close(); } catch (e) { } }
+        }, 4, 30000);
```
Comment (VitGottwald) [inline clients/vscode-hlasmplugin/src/hlasmExternalFilesFtp.ts:64]: named constants for 4 and 30000?

Comment (slavek-kucera) [reply clients/vscode-hlasmplugin/src/hlasmExternalFilesFtp.ts:64]: sure, why not.

File: clients/vscode-hlasmplugin/src/hlasmExternalFilesFtp.ts [hunk 3]
```diff
@@ -46,35 +47,34 @@ class DatasetUriDetails implements ClientUriDetails {
...
+        }, connectionPoolSize, connectionPoolTimeout);
     }
 
-    private clearPoolClient() {
-        clearTimeout(this.pooledClientTimeout);
-        const client = this.pooledClient;
-        this.pooledClient = null;
-        if (client)
-            client.close();
+    get onStateChange() {
```
Comment (VitGottwald) [inline clients/vscode-hlasmplugin/src/hlasmExternalFilesFtp.ts:70]: Is this used anywhere?

Comment (VitGottwald) [reply clients/vscode-hlasmplugin/src/hlasmExternalFilesFtp.ts:70]: ok, found it

File: clients/vscode-hlasmplugin/src/hlasmExternalFilesFtp.ts [hunk 4]
```diff
@@ -106,9 +104,11 @@ export class HLASMExternalFilesFtp implements ExternalFilesClient {
             if (this.activeConnectionInfo)
                 return this.activeConnectionInfo;
 
-            const last = getLastRunConfig(this.context);
+            this.clientPool.closeClients();
+
+            const last = getLastRunConfig(this.context!);
```
Comment (VitGottwald) [inline clients/vscode-hlasmplugin/src/hlasmExternalFilesFtp.ts:109]: How do we lmpw that `this.context` is not null?

Comment (slavek-kucera) [reply clients/vscode-hlasmplugin/src/hlasmExternalFilesFtp.ts:109]: Because it can be undefined only for testing purposes.

**Review by VitGottwald — APPROVED** [review]

**Review by VitGottwald — APPROVED** [review]

File: clients/vscode-hlasmplugin/src/hlasmExternalFilesFtp.ts (hunk 4 above)
Comment (slavek-kucera) [reply clients/vscode-hlasmplugin/src/hlasmExternalFilesFtp.ts:109]: These changes "leaked" from the next PR... I didn't see a reason to not keep them though.

**Pull request merged**